c2corg_api/ext/colander_ext.py

```diff
@@ -53,16 +53,24 @@ def deserialize(self, node, cstruct):
         if cstruct is null or cstruct == '':
             return null
         try:
-            return wkbelement_from_geojson(cstruct, self.srid)
-        except Exception:
-            raise Invalid(node, 'Invalid geometry: %r' % cstruct)
+            geojson = json.loads(cstruct)
+            geom_type = geojson['type'].upper()
+            allowed_types = self.geometry_types
+            if geom_type in allowed_types:
+                return wkbelement_from_geojson(geojson, self.srid)
+            else:
+                raise Invalid(node)
+        except:
+            raise Invalid(
+                node, 'Invalid geometry with expected type: %s. Got %s'
```

Comment:
There are two error cases: The geometry is invalid because of a wrong type or because of a malformed GeoJSON (`wkbelement_from_geojson` fails). Right now, you get the same error message for the two cases. For example, you would get the error message "Invalid geometry with expected type: [POINT]. Got POINT" when sending `{"type": "POINT", some invalid geojson}`. This could be improved. For example:

```python
        try:
            geojson = json.loads(cstruct)
            geom_type = geojson['type'].upper()
            allowed_types = self.geometry_types
            if geom_type in allowed_types:
                return wkbelement_from_geojson(geojson, self.srid)
            else:
                raise Invalid(
                    node, 'Invalid geometry type. Expected: %s. Got: %s.'
                    % (allowed_types, geom_type))
        except:
              raise Invalid(node, 'Invalid geometry: %r' % cstruct)
```